prusti-viper/src/encoder/snapshot/encoder.rs

```diff
@@ -388,12 +401,43 @@ impl SnapshotEncoder {
                 assert_eq!(variants.len(), 1);
                 Ok(variants[0].0.apply(args))
             },
+            Snapshot::Array { cons, .. } => {
+                // the caller must have created a vir::Expr::Seq already
+                assert_eq!(args.len(), 1);
+                assert!(matches!(args[0], vir::Expr::Seq(..)), "Seq expected for array snapshot");
+
+                Ok(cons.apply(args))
+            }
             _ => Err(EncodingError::internal(
                 format!("invalid constructor (not Complex): {}", ty),
             )),
         }
     }
 
+    /// Encode the array indexing operation `a[i]`. Basically translated into `read(snap(a), i)`.
+    pub fn encode_array_idx<'p, 'v: 'p, 'tcx: 'v>(
+        &mut self,
+        encoder: &'p Encoder<'v, 'tcx>,
+        array_ty: ty::Ty<'tcx>,
+        array: vir::Expr,
+        idx: vir::Expr,
+        ) -> EncodingResult<vir::Expr> {
+        let read_func = if let Snapshot::Array { read, .. } = self.encode_snapshot(encoder, array_ty)? {
+            read
+        } else {
+            return Err(EncodingError::internal(
+                format!("called encode_array_idx on non-array-type {:?}", array_ty)
+            ));
+        };
+
+        Ok(read_func.apply(
+            vec![
+                vir::Expr::snap_app(array),
```

Comment:
Given we are in snapshot encoder, can we eagerly expand this `snap_app`? Or is this one of those things that will cause a double borrow?

Reply:
hm, could work if we called `self.patch_snapshots_expr` directly instead of `encoder.patch_snapshots`. i'll test and report back.

Reply:
hm.. doesn't seem to work, but not because double borrows. getting problems with (i think) double applying snapshots(?)
```
error: [Prusti internal error] consistency error: Consistency error: Function snap$__$TY$__Array$3$u32$Snap$Array$3$u32 with formal arguments List(self: Ref) cannot be applied to provided arguments List(cons$Snap$Array$3$u32$__$TY$__Seq$$int$$Snap$Array$3$u32(Seq(0, 1, 2))). (@0.0)
```

FWIW my testcase is `PRUSTI_LOG=prusti_viper::encoder::pure_function_encoder=trace,prusti_viper::encoder::snapshot::encoder=trace ./x.py run --bin prusti-rustc -- --edition 2018 -Pdump_viper_program=true prusti-tests/tests/verify/pass/arrays/array-snap.rs` if you want to have a look yourself.

i think that "duplicate" thing is becase we have a `cons :: Seq[Int] -> Snapshot(Array$...)` and `snap :: TypedRef(Array$...) -> Snapshot(Array$...)`, and even though both declare `Snapshot(Array$...)` as their return type, doing `snap(cons(..))` is a problem.

now that i think about it, we may want to get rid of one of those two, because we're doing the "pure function encoder called from procedure encoder for specifications" case with just `lookup_pure(a, i)` instead anyway, so what's currently called `cons` could just be `snap` for arrays.

Reply:
Hm. `snap_app` (both the `Expr` variant and the `fn` in snapshot encoder) shouldn't do anything when given an already-snapshot thing. What is the input going into `encode_array_idx` here?

Reply:
```
array=Local(_1: Ref(Array$3$u32), Position { line: 0, column: 0, id: 0 }), array.ty=TypedRef("Array$3$u32")
```

Reply:
as the pure function encoder works backwards, it doesn't get the snapshoted thing yet, that happens only later by substituting in the state.

Reply:
Maybe just omit the `snap_app` and see if the patcher fixes this?

Reply:
seems to work, anyway.